### DISCUSSION TITLE
Awesome Pydantic: Projects that use / extend Pydantic in an awesome way

### DISCUSSION THREAD

**Discussion by MartinThoma** [pydantic/pydantic#2503] [General]
I've just seen [flask_pydantic](https://github.com/bauerji/flask_pydantic) and I wonder if anybody has used it before? There seem to be [330 projects which extend pydantic](https://pypi.org/search/?q=pydantic). Can we make an awesomelist with a short description?

## Connecting the Dots

* [flask_pydantic](https://github.com/bauerji/flask_pydantic): validates query and body request parameters and makes them accessible

## Awesome Integration

Making pydantic an integral part of the project:

* [FastAPI](https://pypi.org/project/fastapi/): web framework for building APIs with Python 3.6+ based on standard Python type hints

**Comment by frederikaalund** [comment]
I thank that's a great idea especially now that we have the "settings sources" API (#2107).

**Comment by samuelcolvin** [comment]
This is a great idea.

I'm happy to accept a PR to:
* add an "Awesome Pydantic" page to the docs, it can sit above "Changelog" in the menu
* migrate the "Code Generation" generation into that page
* Add more projects to the page

Each project on the page should have a H1 so it's in the right hand menu, it can then give details about usage and how to install like datamodel-code-generator, or just give a one sentence summary and a link.

The prefix to the page should say:
* "these packages have not been vetted or approved by the pydantic team, your milage may vary" etc.
* "please add your own package here by creating a pull request ..."

**Reply by nuno-andre** [reply]
Given the awesome-lists conventions, maybe an _Awesome Pydantic_ repo would improve visibility, community collaboration, and integration in other awesome-lists. Docs could be easily updated from that repo.

**Reply by MartinThoma** [reply]
@nuno-andre You mean [like this](https://github.com/MartinThoma/awesome-pydantic)?

**Reply by nuno-andre** [reply]
@MartinThoma Just so ;) There's a sort of an awesome-ecosystem with hundreds of awesome-lists that could help to make visible _Pydantic_ and related projects. I think [this awesome-list](https://github.com/sindresorhus/awesome) is the most known source for general topics and [this other](https://github.com/vinta/awesome-python) for Python-related topics.

~~There's also a couple of related GitHub topics: [`awesome`](https://github.com/topics/awesome) and [`awesome-list`](https://github.com/topics/awesome-list) that could be set at repo.~~ Edit: already been done

**Reply by MartinThoma** [reply]
@nuno-andre I've seen many of awesome-* lists. Before I suggest adding the awesome-pydantic to those, I want to have at least 5 projects in there. At the moment I have 2. [I'm looking for more on twitter](https://twitter.com/_martinthoma/status/1377702034552594433), I have seen that there are many on PyPI. Do you know some which you can recommend?

**Reply by nuno-andre** [reply]
Yeah, for sure, I do recommend:

- codegen
  - [`datamodel_code_generator`](https://github.com/koxudaxi/datamodel-code-generator)
    - and I would sublist https://jsontopydantic.com/ as an interesting related tool

- dev toools
  - [the PyCharm plugin](https://github.com/koxudaxi/pydantic-pycharm-plugin)

- projects using pydantic:    
  - I would sublist https://github.com/mjhea0/awesome-fastapi in FastAPI entry (they also list some specifically pydantic-related projects)
  - https://github.com/httprunner/httprunner (declarative http tests, sync)
  - https://github.com/samuelcolvin/arq (job queues, redis, async) maybe @samuelcolvin has something to say about this ;)
  - https://github.com/0b01001001/spectree (api validation)
  - https://github.com/collerek/ormar (orm, async, signaled)
  - https://github.com/art049/odmantic (mongodb odm, async)

I haven't tried those yet, but I think it's worth a look:
- graphql
  - https://github.com/qw-in/graphql-codegen-pydantic
  - https://github.com/lfvilella/pydantic2graphene
- relational
  - https://github.com/drivendataorg/erdantic
  - https://github.com/tiangolo/pydantic-sqlalchemy (from FastAPI author)
- web
  - https://github.com/samuelcolvin/smokeshow (this also brings author's money-back guarantee) 


Also:
- maybe mapping tools (orm, odm, ...) would deserve their own category
- it would be great if you could link both project's repo and docs

**Reply by MartinThoma** [reply]
Thank you! I've added the ones you recommended :-)

**Comment by samuelcolvin** [comment]
Sounds good to me. I'm happy for this to be a separate repo. I'll also link to it from the homepage of the docs and the README.

In terms of comments on the above list:
* arq doesn't make much use of pydantic as far as I can remember
* smokeshow doesn't use pydantic except in tests, its core is written in typescript and the python CLI doesn't depend on pydantic

In terms of popular repos that rely on pydantic, here are a few (with github stars):
* [transformers](https://github.com/huggingface/transformers) - 43k
* fastapi - 29k
* spaCy - 20k
* [ray](https://github.com/ray-project/ray) - 15k
* [ludwig](https://github.com/ludwig-ai/ludwig) from uber - 7.6k
* [DeepPavlov](https://github.com/deepmipt/DeepPavlov) - 5.1k
* [feeluown](https://github.com/feeluown/FeelUOwn) - 2.9k
* [jina](https://github.com/jina-ai/jina) - 2.7k
* [onefuzz](https://github.com/microsoft/onefuzz) from microsoft - 2.2k
* [fastMRI](https://github.com/facebookresearch/fastMRI) from facebook - 600

There are obviously many more, and I know that stars are not a good metric, but that's a start. Please note, I don't know how much all those libraries use pydantic, they're just listed as dependents [in github](https://github.com/samuelcolvin/pydantic/network/dependents) (particularly the packages bit).

**Reply by nuno-andre** [reply]
I didn't know [_ray_ uses _Pydantic_ in Ray Serve](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/ray-project/ray%24+pydantic&patternType=literal). It's indeed a really awesome project.

**Reply by nuno-andre** [reply]
BTW, `sourcegraph.com` could be also a good tool to search for pydantic-related projects. 

A [`lang:python pydantic` search](https://sourcegraph.com/search?q=lang:python+pydantic&patternType=literal) shows projects ordered by number of results. First one is another GraphQL library: https://github.com/strawberry-graphql/strawberry
